hermes/probe/checknil/check_nil.go

```diff
@@ -0,0 +1,332 @@
+func CheckNilFile(ctx context.Context, target *probe.Target, client stiface.Client, logger *logger.Logger) *probe.ProbeError {
+	bucket := target.Target.GetBucketName()
+
+	file, err := checkNilFileExists(ctx, bucket, target, client, logger)
+	if err.Status != metrics.Success {
+		return err
+	}
+
+	if file, err = readNilFile(ctx, file, bucket, target, client, logger); err.Status != metrics.Success {
+		return err
+	}
+
+	if err = verifyNilFile(ctx, file, target, client, logger); err.Status != metrics.Success {
+		return err
+	}
+
+	logger.Infof("CheckNilFile() completed successfully, nil file is valid.")
+	return probe.NewProbeError(metrics.Success, nil)
+}
+
+// checkNilFileExists checks that the nil file exists on the target system.
+// Arguments:
+//	- ctx: context so this probe can be cancelled if needed.
+//	- target: target run information.
+//	- client: initialised storage client for this target system.
+//	- logger: logger associated with the probe calling this function.
+// Returns:
+//	- nilFile: a nilFile struct with the file name recorded
+//	- err: error exit statuses are well defined in metrics.go
+func checkNilFileExists(ctx context.Context, bucket string, target *probe.Target, client stiface.Client, logger *logger.Logger) (*nilFile, *probe.ProbeError) {
+	file := &nilFile{}
+
+	err := RecordAPILatency(target.LatencyMetrics.APICallLatency, metrics.APIListFiles, func() *probe.ProbeError {
+		objIter := client.Bucket(bucket).Objects(ctx, &storage.Query{Prefix: "Hermes_Nil"})
+		for {
+			// TODO(evanSpendlove): Review how Nil file is updated and consider multiple nil files being present.
+			// If one, success
+			// If two -> verify whih one reflects the current state of the system and delete the other.
+			// If two and only differ by intent, prefer the one with intent.
+			obj, err := objIter.Next()
+			if err == iterator.Done {
+				return probe.NewProbeError(metrics.FileMissing, fmt.Errorf("checkNilFileExists(): unable to find nil file: %w", err))
+			}
+			if err != nil {
+				return probe.NewProbeError(metrics.APICallFailed, fmt.Errorf("checkNilFileExists(): unable to list files: %w", err))
+			}
+			file.name = obj.Name
+			return probe.NewProbeError(metrics.Success, nil)
+		}
+	})
+
+	logger.Infof("checkNilFileExists(): nil file located on target storage system.")
+	return file, err
+}
+
+// checkNilFileExists checks that the nil file exists on the target system.
+// Arguments:
+//	- ctx: context so this probe can be cancelled if needed.
+//	- target: target run information.
+//	- client: initialised storage client for this target system.
+//	- logger: logger associated with the probe calling this function.
+// Returns:
+//	- nilFile: a nilFile struct with the file name recorded
+//	- err: error exit statuses are well defined in metrics.go
+func readNilFile(ctx context.Context, file *nilFile, bucket string, target *probe.Target, client stiface.Client, logger *logger.Logger) (*nilFile, *probe.ProbeError) {
```

Comment:
The signature of this function is surprising: it accepts `file` as a pointer, which suggests it's an output argument, but then also returns a `*nilFile`. At the same time, it modifies the input `file` on line 172.
I would refactor this to be a generic `readFile`, which takes a bucket name and a filename and returns a `[]byte`.

Reply:
I've refactored this accordingly.